_drafts/firmware-code-size-optimization-1.md

```diff
@@ -0,0 +1,302 @@
+set:
+
+* `-g`: this generates debug information which puncover requires for analysis
+* `-fdebug-prefix-map=/=`: this flag forces gcc to collect full paths for
+  filenames. This allows puncover to create the hierarchical view of the code.
+Even better would be to use your repository's root (usually found by running
+`git rev-parse --show-toplevel`).
+
+I typically add those two to my CFLAGS for every project directly in my
+Makefile.
+
+### Setting up puncover
+
+Puncover is relatively straightforward to setup: you simply need to clone it,
+setup a virtualenv, and install the dependencies.
+
+```terminal
+# Clone
+francois-mba:code francois$ git clone git@github.com:memfault/puncover.git
+Cloning into 'puncover'...
+remote: Enumerating objects: 1, done.
+remote: Counting objects: 100% (1/1), done.
+remote: Total 609 (delta 0), reused 0 (delta 0), pack-reused 608
+Receiving objects: 100% (609/609), 1.22 MiB | 1.39 MiB/s, done.
+Resolving deltas: 100% (325/325), done.
+# Setup venv
+francois-mba:puncover francois$ python2.7 -m virtualenv venv
+New python executable in /Users/francois/code/puncover/venv/bin/python2.7
+Also creating executable in /Users/francois/code/puncover/venv/bin/python
+Installing setuptools, pip, wheel...
+done.
+# Install dependencies
+francois-mba:puncover francois$ source venv/bin/activate
+(venv) francois-mba:puncover francois$ pip install -r requirements.txt
+[...]
+Successfully installed Flask-0.10.1 Jinja2-2.10.1 MarkupSafe-1.1.1
+Werkzeug-0.15.2 argparse-1.4.0 certifi-2019.3.9 chardet-3.0.4 codecov-2.0.5
+cov-core-1.15.0 coverage-4.5.3 funcsigs-1.0.2 idna-2.8 itsdangerous-1.1.0
+mock-1.3.0 nose-1.3.7 nose-cov-1.6 pbr-5.2.0 requests-2.21.0 six-1.12.0
+urllib3-1.24.3
+```
+
+### Running puncover
+
+While you can install puncover as an executable, I typically just run the
+`runner.py` script directly. All it requires is the path to your GNU ARM
+toolchain:
+
+```terminal
+python runner.py --arm_tools_dir=<path-to-arm-tools-folder> --elf_file <path-to-elf-file>
+```
+
+Note that puncover needs the top level folder for your arm tools, not the `bin`
+file. For me that's `/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/`.
+
+If all goes well, you should see that puncover has spun up a webserver:
+
+```terminal
+(venv) francois-mba:puncover francois$ python runner.py
+--arm_tools_dir=/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/ --e_file
+~/code/zero-to-main/with-libc/build/with-libc.elf
+DEPRECATED: argument --arm_tools_dir will be removed, use --gcc_tools_base
+instead.
+parsing ELF at /Users/francois/code/zero-to-main/with-libc/build/with-libc.elf
+enhancing function sizes
+deriving folders
+enhancing file elements
+enhancing assembly
+enhancing call tree
+enhancing siblings
+unmangling c++ symbols
+ * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit
+```
+
+Now open your favorite web browser at localhost:5000, and enjoy!
+
+Note: You may notice that some code size will be attributed to symbols under
```

Comment:
Specifically the newlib bundled with gcc doesn't build with `-g` fwiw :(

Reply:
Added a note!